### DISCUSSION TITLE
Building firmware with modified platformio.ini

### DISCUSSION THREAD

**Discussion by DaniilMir** [benzino77/tasmocompiler#358] [Q&A]
Hello Community, 

I'm trying to make a Tasmota build with SSL and IR support, but for some IR protocols only.

Original discussion : [https://github.com/arendst/Tasmota/discussions/13202#discussioncomment-6110120](https://github.com/arendst/Tasmota/discussions/13202#discussioncomment-6110120)

I tried connecting to the docker container's shell as per guide : [https://github.com/benzino77/tasmocompiler/blob/master/ADVANCED.md](https://github.com/benzino77/tasmocompiler/blob/master/ADVANCED.md)

After that I've tried adding the following flags: 
`-D_IR_ENABLE_DEFAULT_=false -DDECODE_SYMPHONY=true -DSEND_SYMPHONY=true -DSEND_RAW=true`

into the build_flags section in platformio.ini and platformio_tasmota_env.ini
After running a compile from UI - those settings were lost. 

Questions : 

1. Is there an option to persist these changes in the build_flags? I mean to change flags and to compile a customized firmware.
2. If yes how to do it properly for a generic ESP8266 template ?

**Comment by benzino77** [comment]
To persist changes you need to modify the file [AvailableBoards.js](https://github.com/benzino77/tasmocompiler/blob/master/src/components/AppStepper/FeaturesStep/AvailableBoards.js)
It is rather self describable. I think you need to look at `esp8266generic` board definition and `platformio_entries` section.

**Reply by DaniilMir** [reply]
Thank you @benzino77 for your advise. 
I found this file and modified it as following : 

<details>

<summary>AvailableBoards.js</summary>

>     name: 'esp8266generic',
>     chip_type: 'esp8266',
>     description: 'Generic',
>     default: true,
>     show: true,
>     platformio_entries: {
>       build_flags: -DDECODE_SYMPHONY=true
>                    -DSEND_SYMPHONY=true
>                    -DDECODE_SONY=true
>                    -DSEND_SONY=true
>                    -DSEND_GREE=true
>                    -DDECODE_GREE=true
>                    -DDECODE_ELECTRA_AC=true
>                    -DSEND_ELECTRA_AC=true
>                    -DDECODE_MIRAGE=false
>                    -DDECODE_AIRTON=false
>                    -DDECODE_AIRWELL=false
>                     . . . . . . 
>                    -DSEND_SANYO_AC88=false
>                    -DSEND_SHARP=false
>                    -DSEND_SHARP_AC=false
>                    -DSEND_SHERWOOD=false
>                    -DSEND_TCL112AC=false
>                    -DSEND_TCL96AC=false
>                    -DSEND_TECHNIBEL_AC=false
>                    -DSEND_TECO=false
>                    -DSEND_TEKNOPOINT=false
>                    -DSEND_TOSHIBA_AC=false
>                    -DSEND_TOTO=false
>                    -DSEND_TRANSCOLD=false
>                    -DSEND_TROTEC=false
>                    -DSEND_TROTEC_3550=false
>                    -DSEND_TRUMA=false
>                    -DSEND_VESTEL_AC=false
>                    -DSEND_VOLTAS=false
>                    -DSEND_WHIRLPOOL_AC=false
>                    -DSEND_WHYNTER=false
>                    -DSEND_XMP=false
>                    -DSEND_ZEPEAL=false,
>     },
>     platformio_env_name: 'tasmota',
>     tooltip: 'stepFeaturesBoard8266Tooltip',
>     include_features: ['rules'],

</details>

Unfortunately it didn't work with the resulting build size. 
Maybe I did something wrong with definition. 
The same config added into platformio in the desktop environment works as expected.

**Reply by benzino77** [reply]
`availableBoards` is a **JS object**. You cannot use just literals. You have to close it with single quotes. Something like:

```js
platformio_entries: {
  build_flags: '-DDECODE_SYMPHONY=true -DSEND_SYMPHONY=true -DDECODE_SONY=true -DSEND_SONY=true -DSEND_GREE=true <rest of flags>'
}
```

or it you have to use multi line notation, you should try to use something similar to:
```js
platformio_entries: {
  build_flags:
    '-DDECODE_SYMPHONY=true\n' +
    '\t-DSEND_SYMPHONY=true\n' +
    '\t-DDECODE_SONY=true\n' +
    '\t-DSEND_SONY=true\n' +
    '\t-DSEND_GREE=true\n',
},
```

You can always check how `platformio_override.ini` and ` user_config_override.h` files look like by downloading them after compilation. Those buttons will allow you to download those files to your local machine:
![image](https://github.com/benzino77/tasmocompiler/assets/15658235/27de3d99-9333-4186-85ab-28e384c30065)

You can also check them directly in `/tmp/Tasmota` and `/tmp/Tasmota/tasmota` directory (on gitpod instance)
Then you can review what "magic" TasmoCompiler use to compile your binary (TasmoCompiler tweaks Tasmota by adding things to those two files only - **NOTHING** else ).
